src/qt/updater.cpp

```diff
@@ -0,0 +1,95 @@
+#include "updater.h"
+
+#include <openpgp/hash.h>
+
+#include "utils.h"
+
+Updater::Updater()
+{
+    m_maintainers.emplace_back(fileGetContents(":/monero/utils/gpg_keys/binaryfate.asc").toStdString());
+    m_maintainers.emplace_back(fileGetContents(":/monero/utils/gpg_keys/fluffypony.asc").toStdString());
+    m_maintainers.emplace_back(fileGetContents(":/monero/utils/gpg_keys/luigi1111.asc").toStdString());
+}
+
+QPair<QString, QString> Updater::verifySignaturesAndHashSum(
+    const QByteArray &armoredSignedHashes,
+    const QByteArray &secondDetachedSignature,
+    const QString &binaryFilename,
+    const void *binaryData,
+    size_t binarySize) const
+{
+    QString firstSigner;
+    const QString signedMessage = verifySignature(armoredSignedHashes, firstSigner);
+
+    QString secondSigner = verifySignature(
+        epee::span<const uint8_t>(
+            reinterpret_cast<const uint8_t *>(armoredSignedHashes.data()),
+            armoredSignedHashes.size()),
+        openpgp::signature_rsa::from_buffer(epee::span<const uint8_t>(
+            reinterpret_cast<const uint8_t *>(secondDetachedSignature.data()),
+            secondDetachedSignature.size())));
+
+    if (firstSigner == secondSigner)
+    {
+        throw std::runtime_error("both signatures were generated by the same person");
+    }
+
+    const QByteArray signedHash = parseShasumOutput(signedMessage, binaryFilename);
+    const QByteArray calculatedHash = getHash(binaryData, binarySize);
+    if (signedHash != calculatedHash)
+    {
+        throw std::runtime_error("hash sum mismatch");
+    }
+
+    return {firstSigner, secondSigner};
+}
+
+QByteArray Updater::getHash(const void *data, size_t size) const
+{
+    openpgp::hash hasher(openpgp::hash::algorithm::sha256);
+    hasher << epee::span<const uint8_t>(reinterpret_cast<const uint8_t *>(data), size);
+    const std::vector<uint8_t> hash = hasher.finish();
+    return QByteArray(reinterpret_cast<const char *>(&hash[0]), hash.size());
+}
+
+QByteArray Updater::parseShasumOutput(const QString &message, const QString &filename) const
+{
+    for (const auto &line : message.splitRef("\n"))
+    {
+        if (!line.startsWith(filename))
```

Comment:
Does this match with the current hashes.txt format? Shouldn’t the line start with the hash, then double space and then the filename?

I added my key to the source, detached signed the hashes.txt and tried it out:
```
selsta@mbpR ~/d/monero-gui (2824) [1]> build/release/bin/monero-wallet-gui.app/Contents/MacOS/monero-wallet-gui --verify-update ~/Downloads/monero-gui-mac-x64-v0.15.0.4.tar.bz2
2020-04-08 00:03:57.244	W app startd (log: /Users/selsta/Library/Logs/monero-wallet-gui.log)
2020-04-08 00:03:57.261	E hash not found
```

Reply:
Fixed. Supports both formats now.